pypsa/optimization/abstract.py

```diff
@@ -806,42 +1000,225 @@ def optimize_mga(
             raise ValueError(msg)
 
         # build alternate objective
-        objective = []
-        for c, attrs in weights.items():
-            for attr, coeffs in attrs.items():
-                if isinstance(coeffs, dict):
-                    coeffs = pd.Series(coeffs)
-                if attr == nominal_attrs[c] and isinstance(coeffs, pd.Series):
-                    coeffs = coeffs.reindex(self._n.get_extendable_i(c))
-                    coeffs.index.name = ""
-                elif isinstance(coeffs, pd.Series):
-                    coeffs = coeffs.reindex(columns=self._n.static(c).index)
-                elif isinstance(coeffs, pd.DataFrame):
-                    coeffs = coeffs.reindex(
-                        columns=self._n.static(c).index, index=snapshots
-                    )
-                objective.append(m[f"{c}-{attr}"] * coeffs * sense)
-
-        m.objective = merge(objective)
+        m.objective = self.build_linexpr_from_weights(weights, model=m) * sense
 
         status, condition = self._n.optimize.solve_model(**kwargs)
 
         # write MGA coefficients into metadata
         self._n.meta["slack"] = slack
         self._n.meta["sense"] = sense
+        self._n.meta["weights"] = _convert_to_dict(weights)
 
-        def convert_to_dict(obj: Any) -> Any:
-            if isinstance(obj, pd.DataFrame):
-                return obj.to_dict(orient="list")
-            if isinstance(obj, pd.Series):
-                return obj.to_dict()
-            if isinstance(obj, dict):
-                return {k: convert_to_dict(v) for k, v in obj.items()}
-            return obj
+        return status, condition
 
-        self._n.meta["weights"] = convert_to_dict(weights)
+    def project_solved(
```

Comment:
Same here, let's make it internal if the user should not use it directly

```suggestion
    def _project_solved(
```